jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/openshift/ImageStreamService.java

```diff
@@ -172,59 +172,81 @@ private TagReference extractTag(ImageStream is) {
-            for (NamedTagEventList list : tags) {
-                List<TagEvent> items = list.getItems();
-                if (items == null || items.isEmpty()) {
-                    continue TAG_EVENT_LIST;
+                if (latestTag != null && StringUtils.isNotBlank(latestTag.getImage())) {
+                    String image = latestTag.getImage();
+                    log.info("Found tag on ImageStream " + imageStreamName + " tag: " + image);
+                    return image;
                 }
+            }
+        }
+        // Handle the case when no image is found after all retries:
+        throw handleNoImageFoundException(currentImageStream, imageStreamName);
+    }
 
-                for (TagEvent tag : items) {
-                    latestTag = latestTag == null ? tag : newerTag(tag, latestTag);
-                }
+    // Method to sleep the current thread for a given amount of time
+    private void sleepThread() {
+        try {
+            Thread.sleep(IMAGE_STREAM_TAG_RETRY_TIMEOUT_IN_MILLIS);
+        } catch (InterruptedException e) {
+            log.debug("interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    // Method to find the latest TagEvent from an ImageStream
+    private TagEvent findLatestTag(ImageStream currentImageStream) {
+        ImageStreamStatus status = currentImageStream.getStatus();
+        if (status != null) {
+            List<NamedTagEventList> tags = status.getTags();
+            if (tags != null && !tags.isEmpty()) {
+                return getLatestTagFromEventLists(tags);
             }
+        }
+        return null;
+    }
 
-            if (latestTag != null && StringUtils.isNotBlank(latestTag.getImage())) {
-                String image = latestTag.getImage();
-                log.info("Found tag on ImageStream " + imageStreamName + " tag: " + image);
-                return image;
+    // Method to iterate over NamedTagEventList and find the latest TagEvent
+    private TagEvent getLatestTagFromEventLists(List<NamedTagEventList> tags) {
+        TagEvent latestTag = null;
+        for (NamedTagEventList list : tags) {
+            latestTag = updateLatestTagFromList(latestTag, list);
+        }
+        return latestTag;
+    }
+
+    // Method to find the latest TagEvent from a NamedTagEventList
+    private TagEvent updateLatestTagFromList(TagEvent latestTag, NamedTagEventList list) {
+        List<TagEvent> items = list.getItems();
+        if (items != null && !items.isEmpty()) {
+            for (TagEvent tag : items) {
+                // Update the latest tag
+                latestTag = latestTag == null ? tag : newerTag(tag, latestTag);
             }
         }
+        //
+        return latestTag;
+    }
 
-        // No image found, even after several retries:
+    // Method to handle cases where no image is found
+    private IllegalStateException handleNoImageFoundException(ImageStream currentImageStream, String imageStreamName) {
```

Comment:
```suggestion
    private String handleNoImageFoundException(ImageStream currentImageStream, String imageStreamName) {
```
Maybe instead of returning exception, you can get return exception message.